library/src/main/java/org/apache/apex/malhar/lib/state/managed/ManagedTimeStateMultiValue.java

```diff
@@ -0,0 +1,357 @@
+   * @param k key
+   * @param v value
+   * @param timeBucket timebucket
+   * @return true if the given (k,v) is successfully inserted into the store otherwise false.
+   */
+  public boolean put(@Nullable K k, @Nullable V v, long timeBucket)
+  {
+    if (isKeyContainsMultiValue) {
+      Slice keySlice = streamCodec.toByteArray(k);
+      int bucketId = getBucketId(k);
+      Slice valueSlice = store.getSync(bucketId, keySlice);
+      List<V> listOb;
+      if (valueSlice == null || valueSlice.length == 0) {
+        listOb = new ArrayList<>();
+      } else {
+        listOb = (List<V>)streamCodec.fromByteArray(valueSlice);
+      }
+      listOb.add(v);
+      return insertInStore(bucketId, timeBucket, keySlice, streamCodec.toByteArray(listOb));
+    }
+    return insertInStore(getBucketId(k), timeBucket, streamCodec.toByteArray(k),streamCodec.toByteArray(v));
+  }
+
+  /**
+   * Insert (keySlice,valueSlice) into the store using bucketId and timeBucket.
+   * @param bucketId bucket Id
+   * @param timeBucket time bucket
+   * @param keySlice key slice
+   * @param valueSlice value slice
+   * @return true if the given (keySlice,valueSlice) is successfully inserted into the
+   *         store otherwise false.
+   */
+  public boolean insertInStore(long bucketId, long timeBucket, Slice keySlice, Slice valueSlice)
+  {
+    long timeBucketId = store.getTimeBucketAssigner().getTimeBucketFor(timeBucket);
+    if (timeBucketId != -1) {
+      store.putInBucket(bucketId, timeBucketId, keySlice, valueSlice);
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public boolean remove(@Nullable Object o, @Nullable Object o1)
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean putAll(@Nullable K k, Iterable<? extends V> iterable)
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public boolean putAll(Multimap<? extends K, ? extends V> multimap)
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public List<V> replaceValues(K k, Iterable<? extends V> iterable)
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Map<K, Collection<V>> asMap()
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  public int getBucketId(K k)
+  {
+    return k.hashCode() % store.getNumBuckets();
+  }
+
+
+  public boolean isKeyContainsMultiValue()
+  {
+    return isKeyContainsMultiValue;
+  }
+
+  public void setKeyContainsMultiValue(boolean keyContainsMultiValue)
+  {
+    isKeyContainsMultiValue = keyContainsMultiValue;
+  }
+
+  public long getTimeBucket()
+  {
+    return timeBucket;
+  }
+
+  public void setTimeBucket(long timeBucket)
+  {
+    this.timeBucket = timeBucket;
+  }
+
+  public StreamCodec getStreamCodec()
+  {
+    return streamCodec;
+  }
+
+  public void setStreamCodec(StreamCodec streamCodec)
+  {
+    this.streamCodec = streamCodec;
+  }
+
+  public class CompositeFuture implements Future<List>
```

Comment:
Instead of having this entire class, is it possible that you can just create a call called ```getAsync()```? You can then return the slice as a ```Future``` and let the caller handle whether the contents of the slice is a single value or multi values.